examples/graph/gat_node_classification.py

```diff
@@ -0,0 +1,402 @@
+        return tf.nn.relu(outputs)
+
+
+def GraphAttentionNetwork(hidden_units, num_heads, num_layers):
+
+    node_features = layers.Input((NODE_DIM), dtype="float32")
+    adjacency = layers.Input((2), dtype="int32")
+
+    x = layers.Dense(hidden_units * num_heads, activation="relu")(node_features)
+
+    for _ in range(num_layers):
+        x_updated = MultiHeadGraphAttention(hidden_units, num_heads)([x, adjacency])
+        x = layers.Add()([x_updated, x])
+
+    out = layers.Dense(OUTPUT_DIM)(x)
+
+    model = keras.Model(inputs=[node_features, adjacency], outputs=[out])
+
+    return model
+
+
+# Build graph attention network model
+gat_model = GraphAttentionNetwork(HIDDEN_UNITS, NUM_HEADS, NUM_LAYERS)
+
+# Visualize built model
+keras.utils.plot_model(gat_model, show_shapes=True, show_layer_names=False)
+
+"""
+### Define custom `train_step`
+"""
+
+
+def get_train_step_fn(model, loss_fn, optimizer):
+    @tf.function
+    def train_step(inputs, labels, indices):
+        with tf.GradientTape() as tape:
+            outputs = model(inputs)
+            loss = loss_fn(labels, tf.gather(outputs, indices))
+        grads = tape.gradient(loss, model.trainable_weights)
+        optimizer.apply_gradients(zip(grads, model.trainable_weights))
+        return loss, outputs
+
+    return train_step
+
+
+"""
+### Train and predict
+"""
+
+loss_fn = keras.losses.SparseCategoricalCrossentropy(from_logits=True)
+optimizer = keras.optimizers.SGD(LEARNING_RATE, momentum=MOMENTUM)
+accuracy_fn = lambda y_true, y_pred: tf.reduce_mean(
+    keras.metrics.sparse_categorical_accuracy(y_true, y_pred)
+)
+
+train_step = get_train_step_fn(gat_model, loss_fn, optimizer)
+
+best_accuracy = 0.0
+for i in range(NUM_ITERATIONS):
+
+    loss, preds = train_step(
+        inputs=(node_features, adjacency), labels=train_labels, indices=train_indices
+    )
+
+    # Obtain valid accuracy, and if best so far, obtain test accuracy
+    # and test probabilities (for later use)
+    valid_preds = tf.gather(preds, valid_indices)
+    valid_accuracy = accuracy_fn(valid_labels, valid_preds)
+    if valid_accuracy > best_accuracy:
+        best_accuracy = valid_accuracy
+        test_preds = tf.gather(preds, test_indices)
+        test_accuracy = accuracy_fn(test_labels, test_preds)
+        test_probs = tf.nn.softmax(test_preds).numpy()
+
+    if i % 10 == 0:
+        print(
+            f"Iteration {i:03d} : "
+            + f"Loss {loss:.3f} : "
+            + f"Valid Accuracy {valid_accuracy*100:.1f}%"
+        )
+
+
+print("--" * 26)
+print(f"\t\t\t  Test Accuracy {test_accuracy*100:.1f}%")
+
+"""
+### Inspect predictions
+"""
+
+mapping = {v: k for (k, v) in class_idx.items()}
+
+for i, (probs, label) in enumerate(zip(test_probs[:10], test_labels[:10])):
+    print(f"Example {i+1}: {mapping[label]}")
+    for j, c in zip(probs, class_idx.keys()):
+        print(f"\tProbability of {c: <24} = {j*100:7.3f}%")
+    print("---" * 20)
+
+"""
+**Not too bad!**
```

Comment:
Please comment in more detail what the results mean and how we could further improve them.